Source/Core/Core/IOS/USB/Bluetooth/LibUSBBluetoothAdapter.cpp

```diff
@@ -0,0 +1,570 @@
+// Copyright 2025 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/IOS/USB/Bluetooth/LibUSBBluetoothAdapter.h"
+
+#include <algorithm>
+#include <cstdlib>
+#include <cstring>
+#include <ranges>
+#include <utility>
+
+#include <fmt/format.h>
+#include <libusb.h>
+
+#include "Common/MsgHandler.h"
+
+#include "Core/Config/MainSettings.h"
+#include "Core/Core.h"
+#include "Core/IOS/USB/Bluetooth/hci.h"
+#include "Core/IOS/USB/Host.h"
+
+namespace
+{
+constexpr std::size_t BUFFER_SIZE = 1024;
+
+constexpr auto HCI_COMMAND_TIMEOUT = std::chrono::milliseconds{1000};
+
+constexpr u8 REQUEST_TYPE =
+    u8(u8(LIBUSB_ENDPOINT_OUT) | u8(LIBUSB_REQUEST_TYPE_CLASS)) | u8(LIBUSB_RECIPIENT_DEVICE);
+
+constexpr u8 HCI_EVENT = 0x81;
+constexpr u8 ACL_DATA_IN = 0x82;
+
+template <auto MemFun>
+constexpr libusb_transfer_cb_fn LibUSBMemFunCallback()
+{
+  return [](libusb_transfer* tr) {
+    std::invoke(MemFun, static_cast<Common::ObjectType<MemFun>*>(tr->user_data), tr);
+  };
+}
+
+bool IsBluetoothDevice(const libusb_device_descriptor& descriptor)
+{
+  constexpr u8 SUBCLASS = 0x01;
+  constexpr u8 PROTOCOL_BLUETOOTH = 0x01;
+  // Some devices misreport their class, so we avoid relying solely on descriptor checks and allow
+  // users to specify their own VID/PID.
+  return LibUSBBluetoothAdapter::IsConfiguredBluetoothDevice(descriptor.idVendor,
+                                                             descriptor.idProduct) ||
+         (descriptor.bDeviceClass == LIBUSB_CLASS_WIRELESS &&
```

Comment:
These checks would be cheaper than querying config, so maybe they should come first?

Also, maybe:

```
const bool is_bluetooth_protocol = descriptor.bDeviceClass == LIBUSB_CLASS_WIRELESS && descriptor.bDeviceSubClass == SUBCLASS && descriptor.bDeviceProtocol == PROTOCOL_BLUETOOTH;

return is_bluetooth_protocol || LibUSBBluetoothAdapter::IsConfiguredBluetoothDevice(descriptor.idVendor, descriptor.idProduct);
```
would be more readable?